irq: esp32: Interrupts priority configuration

raffarost opened the pull request:
Allows configuring interrupts priority and flags in the device tree for Espressif SoC's

Commit:
drivers: esp32: esp_intr_alloc return condition

Add checks to return value of esp_intr_alloc to avoid drivers init
returning 0 when interrupt allocation fails.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

marekmatej reviewed (commented):
In general LGTM, just consider priority vs level terminology, so we wont introduce new terms here which diverge from the TRM

raffarost commented on drivers/can/can_esp32_twai.c:
```diff
@@ -71,6 +71,7 @@ struct can_esp32_twai_config {
 	const struct device *clock_dev;
 	const clock_control_subsys_t clock_subsys;
 	int irq_source;
+	int irq_priority;
```
Picking priority vs level is an effort to steer the DTS in the direction of Zephyr's interrupt system.
Source was unchanged because it would be explicitly wrong to call it irq_number.
We can discuss and change it though, if you disagree.

wmrsouza replied:
I seem's that irq_priority is more present along the Zephyr source code.

Commit:
soc: mp: esp32: Added IRQ priority and flags config

Interrupt priority and flags config for crosscore ISR
sourced from device tree.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

marekmatej commented on drivers/can/can_esp32_twai.c:
```diff
@@ -202,9 +204,16 @@ static int can_esp32_twai_init(const struct device *dev)
-	esp_intr_alloc(twai_config->irq_source, 0, can_esp32_twai_isr, (void *)dev, NULL);
+	err = esp_intr_alloc(twai_config->irq_source,
+			esp_intr_level_to_flags(twai_config->irq_priority) |
+			esp_intr_flags_check(twai_config->irq_flags) | ESP_INTR_FLAG_IRAM,
```
Maybe a macro creating flags out of the `irq_priority` and `irq_flags` would help in usability.

raffarost replied:
Inline functions were replaced by macros

Commit:
bindings: intc: esp32: Added symbols in DT

Added symbols for IRQ priority and flags configuration
in the device tree.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
dts: irq: esp32: Added priority and flags to device tree

Added IRQ priority and flags configuration to device tree.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
drivers: intc: esp32: Macros for DT input validation

Adding macros to convert and validate input config from device tree
for IRQ priority and flags.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
drivers: esp32: Interrupts priority configuration

Allows configuring interrupts priority in the device tree for
ESP32 devices.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
drivers: esp32: Interrupts flags configuration

Allows configuring interrupts flags in the device tree for
ESP32 devices.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>
